### ISSUE TITLE
Request: Plugin PES cleaner

### ISSUE THREAD

**Issue by lars18th** [tsduck/tsduck#273]
Hi,

I observed that some streams use PADDING inside the PES HEADER, instead of inside the TS HEADER.

This is a good idea if the entire PES data needs to be in a deterministic number of TS packets. Then is possible to remux, and the number of TS packets doesn't change. However, when the packet is remuxed but, for example, a PCR is inserted, then the length alignement is broken.

So, I ask if it's possible to develop a simple plugin called `PES cleaner` that will reduce the PES Header Padding (0xFF) and replaces it with NULL stuff in the TS Header. This can be done at packet level (inside the same packet), and the plugin can use the list of pids to process as a parameter.

What you think?

**Comment by lars18th** [comment]
Hi @lelegard ,

I try to explain this request with one example. See this packet of TELETEXT data:

```
         0001 0203 0405 0607 0809 0A0B 0C0D 0E0F 0123456789ABCDEF
0x000000 4744 4E14 0000 01BD 02DA 8480 2421 B803 GDN.........$!..
0x000010 F069 FFFF FFFF FFFF FFFF FFFF FFFF FFFF .i..............
0x000020 FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF ................
0x000030 FF10 022C E7E4 F46D A880 C900 FC92 932E ...,...m........
0x000040 FEFF 2EFE FF2E FEFF 2EFE FF2E FEFF 2EFE ................
0x000050 FF2E FEFF 2EFE FF2E FEFF 2EFE FF86 4200 ..............B.
0x000060 022C E8E4 576D A857 57A8 A8A8 A857 57A8 .,..Wm.WW....WW.
0x000070 A8A8 A857 57A8 A8A8 A857 57A8 A8A8 A857 ...WW....WW....W
0x000080 5757 F457 7AA8 A857 F457 570B 1D48 022C WW.Wz..W.WW..H.,
0x000090 E9E4 57A8 E9B9 0404 040E 0E0D 0404 040E ..W.............
0x0000a0 0404 040D 040E 0E04 040D 040E 0E04 040E ................
0x0000b0 0404 040D 0404 0E0D 040D 0404           ............
```

This is the first packet of a PES sequence.
As you can see the PES length is 730 bytes (0x02DA @0x08, so 3 more packets come after), and the PES header has a size of 36 bytes (0x24 @0x0C). So it has padding stuff in the PES header. The other 3 packets have full payload (184 byes), including the last as it ends at boundires. The PES bytes used by the four packets are then: 142 + 184 + 184 + 184.

But my objective is to remove the unused 31 bytes of the PES padding and convert it to TS padding with adaptation field. This process can be done at packet level, as all required information is available inside the same packet (in fact only PES header length & PES length). And we don't need to move any data, but patch the headers.

After this modification it will be possible to change the packet, for example adding PCR marks, without touching the PES content.

You agree with this functionality?

**Comment by lars18th** [comment]
I close this issue as the `--pack-pes-header` it's now implemented in the filter `craft` (#291).
